gateway/src/management_switch.rs

```diff
@@ -203,14 +203,28 @@ impl ManagementSwitch {
         self.location_map().map(|m| m.location_name())
     }
 
-    /// Get the socket to use to communicate with an SP and the socket address
-    /// of that SP.
-    pub(crate) fn sp(&self, port: SwitchPort) -> Option<&SingleSp> {
-        self.sockets.get(&port)
+    /// Get the handle for communicating with an SP by its switch port.
+    pub(crate) fn sp(&self, port: SwitchPort) -> &SingleSp {
+        // `SwitchPort` is an opaque type; we're guaranteed to have an entry for
+        // every port.
+        self.sockets.get(&port).unwrap()
+    }
+
+    /// Get an iterator providing the ID and handle to communicate with every SP
+    /// we know about.
+    ///
+    /// This function can only fail if we have not yet completed discovery (and
+    /// therefore can't map our switch ports to SP identities).
+    pub(crate) fn all_sps(
```

Comment:
This method is the change that I find maybe the most confusing/dubious in this PR: this returns an iterator of all SPs we _potentially_ know about (i.e., every switch port we're configured to use to communicate with an SP). In practice on real hardware this will always return an iterator with 35 items, even if some of the slots are currently unoccupied. This means that if any slot is unoccupied, the "get all SP state" endpoint will not return faster than it takes for the requests to that slot to timeout. This is maybe a bigger conversation than just this PR, but I'd like to address this soon in a followup.

Reply:
I think the question here is "How do we know when we shouldn't check all SPs?". Unfortunately, I also think the answer is we can never know, since we can't rely on ignition to tell us. Given that, I don't see how we workaround this. In most cases we expect racks to be full, but there will always be failures and upgrades, and we will timeout some large percentage of the the time. In that case, I think we may have to just accept that and make the timeout as short as possible.

Reply:
Fundamentally I agree with you; however, if we have a permanently (or effectively permanently) underpopulated rack, it's very unfortunate that this will always time out. I have similar concerns about detecting removal / replacement / swapping of a gimlet; that is definitely not handled correctly right now.

Thinking out loud, maybe each `SingleSp` should remember the last time it heard from the SP it thinks it should be talking to (which in general should be less than the interval on which we send periodic "discover" packets, which is currently < 60 seconds), and if that time is sufficiently long + ignition says there's no gimlet present we stop waiting for it?

Reply:
That sounds reasonable, but only if there is some guaranteed communication from the SP.